File: fsharp/ch2/simple-math/Input.fs
module SimpleMath.Input

open System
open SimpleMath.Core

let private makeParseState (success:bool,result:float) =
    match success with
    | true -> if result < 0.00 then Negative else Success(result)
    | false -> Failure

let parse (str:string) =
    Double.TryParse str
    |> makeParseState

File: clojure/ch2/simple-math/src/brianium/simple_math/input.clj
(ns brianium.simple-math.input
  (:require [clojure.spec.alpha :as s]
            [brianium.simple-math.input.spec :as input]))

(defn make-parse-state
  [f]
  (if (< f 0)
    {:state :negative :value f}
    {:state :success :value f}))

(s/fdef make-parse-state
  :args (s/cat :f float?)
  :ret  ::input/state)

(defn parse
  [s]
  (try
    (make-parse-state (Float/parseFloat s))
    (catch Exception _
      {:state :failure})))

(s/fdef parse
  :args (s/cat :s string?)
  :ret  (s/tuple float? boolean?))

(defn success?
  [parse-state]
  (= :success (:state parse-state)))

(s/fdef success?
  :args (s/cat :state ::input/state)
  :ret  boolean?)
